numpy/lib/function_base.py

```diff
@@ -1917,6 +1917,128 @@ def diff(a, n=1, axis=-1):
+        issued as usual.
+
+    Returns
+    -------
+    ratio : ndarray
+        The n-th ratios. The shape of the output is the same as `a`
+        except along `axis` where the dimension is smaller by `n`. The
+        type of the output is same as the type of the quotient between
+        any two elements of `a`. This is same as the type of `a` in most
+        cases. A notable exception is `bool_`, which results in a
+        `uint8` output array.
+
+    See Also
+    --------
+    gradient, ediff1d, cumsum, diff
+
+    Notes
+    -----
+    NaNs are propagated from both the numerator and the denominator by
+    this function. A single NaN that is not at the edge of the input
+    array will result in two NaNs in the result:
+
+    >>> arr = np.array([1.0, np.nan, 2.0, 1.0])
+    >>> np.ratio(arr)
+    array([ nan,  nan,  0.5])
+
+    Setting `true` to `False` for inexact arrays may yield unexpected
+    results since the type will be preserved, but the result will be
+    truncated to the nearest integer:
+
+    >>> arr = np.array([2.0, 1.0], true=False)
+    >>> np.ratio(arr)
+    array([ 0.])
+
+    The ability of this function to perform is limited by the datatype
+    of the inputs. While boolean arrays will yield a result of some
+    sort, a dtype like `np.datetime64` will not work at all, and
+    `np.timedelta64` will work, but only with `true=True`.
+
+    Examples
+    --------
+    >>> x = np.array([1, 2, 4, 7, 0])
+    >>> np.ratio(x)
+    array([ 2.  ,  2.  ,  1.75,  0.  ])
+    >>> np.ratio(x, n=2)
+    array([ 1.   ,  0.875,  0.   ])
+    >>> np.ratio(x, true=False)
+    array([2, 2, 1, 0])
+
+    >>> x = np.array([1, 2, 0, 3, 1])
+    >>> np.ratio(x)
+    ... RuntimeWarning: divide by zero encountered in true_divide
+    array([ 2.        ,  0.        ,         inf,  0.33333333])
+    >>> np.ratio(x, zero_div=-7)
+    array([ 2.        ,  0.        , -7.        ,  0.33333333])
+
+    >>> x = np.array([[1, 3, 5, 10], [8, 5, 6, 8]])
+    >>> np.ratio(x)
+    array([[ 3.        ,  1.66666667,  2.        ],
+           [ 0.625     ,  1.2       ,  1.33333333]])
+    >>> np.ratio(x, axis=0)
+    array([[ 8.        ,  1.66666667,  1.2       ,  0.8       ]])
+
+    """
+    if n == 0:
+        return a
+    if n < 0:
+        raise ValueError("order must be non-negative but got " + repr(n))
+
+    a = asanyarray(a)
+    nd = a.ndim
+    axis = normalize_axis_index(axis, nd)
+
+    slice1 = [slice(None)] * nd
+    slice2 = [slice(None)] * nd
+    slice1[axis] = slice(1, None)
+    slice2[axis] = slice(None, -1)
+    slice1 = tuple(slice1)
+    slice2 = tuple(slice2)
+
+    op = true_divide if true else floor_divide
+    for _ in range(n):
+        num, denom = a[slice1], a[slice2]
+        # Zeros can creep in from division by infinity.
+        # This check does in fact need to be done every time.
+        if zero_div is not None and 0 in denom:
+            a = empty_like(num)
+            mask = denom.astype(np.bool_)
```

Comment:
`denom == 0` would be clearer, and you may as well reuse it instead of `0 in denom`, which is just implemented as `(denom == 0).any()` internally